gnomad/utils/constraint.py

```diff
@@ -563,6 +568,247 @@ def collapse_strand(
+    )
+
+    # Add annotations for 'ref' and 'alt'.
+    ht = ht.annotate(ref=ht.alleles[0], alt=ht.alleles[1])
+
+    # Trim heptamer context to create trimer context and filter to where the bases are
+    # either A, T, C, or G.
+    ht = trimer_from_heptamer(ht)
+    ht = ht.filter(ht.context.matches(f"[ATCG]{{{3}}}"))
+
+    # Annotate mutation type (such as "CpG", "non-CpG transition", "transversion") and
+    # collapse strands to deduplicate the context.
+    ht = annotate_mutation_type(collapse_strand(ht))
+
+    # Add most_severe_consequence annotation to 'transcript_consequences' within the
+    # vep root annotation.
+    csqs = ht.vep.transcript_consequences
+    csqs = add_most_severe_consequence_to_consequence(csqs)
+    vep_csq_fields = [
+        "transcript_id",
+        "gene_id",
+        "gene_symbol",
+        "biotype",
+        "most_severe_consequence",
+        "mane_select",
+        "canonical",
+        "lof",
+        "lof_flags",
+    ]
+    vep_csq_fields = [x for x in vep_csq_fields if x in csqs.dtype.element_type]
+    ht = ht.annotate(
+        vep=ht.vep.select(
+            "most_severe_consequence",
+            transcript_consequences=csqs.map(lambda x: x.select(*vep_csq_fields)),
+        )
+    )
+
+    # Add 'methylation_level', 'coverage', and 'gerp' annotations if specified.
+    transformation_funcs = transformation_funcs or {}
+
+    if "methylation_level" not in transformation_funcs:
+        transformation_funcs["methylation_level"] = lambda x, t: hl.if_else(
+            ht.cpg, trans_methyl(x)[t.locus].methylation_level, 0
+        )
+
+    if "gerp" not in transformation_funcs:
+        transformation_funcs["gerp"] = lambda x, t: hl.if_else(
+            hl.is_missing(x[t.locus].S), 0, x[t.locus].S
+        )
+
+    # If necessary, pull out first element of coverage statistics (which includes all
+    # samples). Relevant to v4, where coverage stats include additional elements to
+    # stratify by ukb subset and platforms.
+    if "coverage" not in transformation_funcs:
+        transformation_funcs["coverage"] = lambda x, t: (
+            x[t.locus].coverage_stats[0] if "coverage_stats" in x.row else x[t.locus]
+        )
+
+    if "AN" not in transformation_funcs:
+        transformation_funcs["AN"] = lambda x, t: x[t.locus].AN
+
+    if "freq" not in transformation_funcs:
+        transformation_funcs["freq"] = lambda x, t: x[t.key].freq
+
+    if "filters" not in transformation_funcs:
+        transformation_funcs["filters"] = lambda x, t: x[t.key].filters
+
+    hts = {
+        "methylation_level": methylation_ht,
+        "gerp": gerp_ht,
+        "coverage": coverage_hts,
+        "AN": an_hts,
+        "freq": freq_hts,
+        "filters": filter_hts,
+    }
+    hts = {k: v for k, v in hts.items() if v is not None}
+    exprs = {}
+    for ann, ann_ht in hts.items():
+        if isinstance(ann_ht, dict):
+            ann_expr = hl.struct(
```

Comment:
Can you share me the flowchart how to calculate constraint? I don't understand why you need to have coverage, AN, freq, and filters from both exomes and genomes.

Reply:
Part of it is just that it's nice to have it all in one table, and then they user can decide which metric (coverage or AN) to use. The idea for this HT is to remove the need for processing the context, exomes, and genomes HTs individually. If it's all in the same HT, you only need to process tha HT once.

I don't currently have the PRs ready for all the other changes I have made to get the pipeline simplified, this is a first step. I will put those in later (hopefully soon) and update the constraint flowchart when finished